### PR TITLE
Gh-275: Schema Structure documentation

### LINKED ISSUE #275: schema structure documentation needs review
The "getting started" documentation recommends:

"The Schema file can be broken down into small parts, we encourage at least 2 files:

[elements.json](https://github.com/gchq/gaffer-doc/blob/master/user/src/main/resources/RoadUse/schema/elements.json)
[types.json](https://github.com/gchq/gaffer-doc/blob/master/user/src/main/resources/RoadUse/schema/types.json)
Splitting the schema up into these 2 files helps to illustrate the different roles that the schemas fulfil.
"

We don't do this. Is this advice out of date? Perhaps it should say that the schema can be split across multiple files? What naming conventions apply? Can they be overridden? etc

### PR DESCRIPTION
**Pull request by GCHQDeveloper314**
# Related Issue

- Resolve #275

### CODE REVIEW

File: docs/getting-started/advanced-guide/schema.md [hunk 1]
```diff
@@ -0,0 +1,429 @@
...
+Entities describe a vertex. We use the term "element" to mean either and edge or and entity.
+
+When defining an element we must provide a "group". This is a unique string identifier for each element.
+Groups must be completely unique and cannot be shared between edges and entities.
+
+### Edges
+Edges must have the following:
+
+- `source` - The type of object that will be used as the source vertex in your graph. Needs to be a type defined in the [Types schema](#types).
+- `destination` - Similar to source, it can either be the same or a different type.
```
Comment (cn337131) [inline docs/getting-started/advanced-guide/schema.md]: Potentially should mention that this is the destination vertex, just to ensure clarity.

File: docs/getting-started/advanced-guide/schema.md [hunk 2]
```diff
@@ -0,0 +1,429 @@
...
+- `directed` - Tells Gaffer if the edge is directed or undirected. Needs to be a type which means true or false, see [Types](#true--false) for more info.
+
+When an Edge is undirected in Gaffer (`directed` is `false`), it is treated as if the relationship was bidirectional and the vertices of the edge do not have an authoritative source and destination thus, the undirected edges `A -- B` and `B -- A` are equal, and therefore will be aggregated together.
```
Comment (cn337131) [inline docs/getting-started/advanced-guide/schema.md]:
```suggestion
When an Edge is undirected in Gaffer (`directed` is `false`), it is treated as if the relationship was bidirectional and the vertices of the edge do not have an authoritative source and destination. Thus, the undirected edges `A -- B` and `B -- A` are equal, and therefore will be aggregated together.
```

File: docs/getting-started/advanced-guide/schema.md [hunk 3]
```diff
@@ -0,0 +1,429 @@
...
+Gaffer will present the undirected edges Vertices in natural ordering, so a client will also see the above edge presented as `A, B`.
+It will be aggregated with any other edge that has the same source and destination and is also undirected.
+For example an undirected Edge that is added, `A -> B`, will be aggregated with another Edge of `B -> A` if they are undirected.
+
+### Entities
+Entities must have a `vertex` field, which is similar to the `source` and `destination` fields on an edge.
+For example, modelling `London -> Paris` requires an entity definition to represent the city and an edge definition to represent the relationship.
+In that example the entity `vertex` field and the edge `source` and `destination` fields would all be of the same type.
```
Comment (cn337131) [inline docs/getting-started/advanced-guide/schema.md]:
```suggestion
In this example the entity `vertex` field and the edge `source` and `destination` fields would all be of the same type.
```

File: docs/getting-started/advanced-guide/schema.md [hunk 4]
```diff
@@ -0,0 +1,429 @@
...
+- `parents` - An array of parent group names. These must relate to the same sort of element as the child, for example an edge cannot have an entity as a parent. Elements can inherit any information from multiple parent elements. Fields will be merged/overridden, so the hierarchy of parents is important. Any fields that are defined in the child element will also merge or override information taken from the parents.
+- `properties` - Properties are defined by a map of key-value pairs of property names to property types. Property types are described in the Types schema.
+- `groupBy` - By default Gaffer uses the element group and its vertices to group similar elements together when aggregating and summarising elements. This groupBy field allows you to specify extra properties (in addition to the element group and vertices) to use for controlling when similar elements should be grouped together and summarised.
+- `visibilityProperty` - When using visibility properties in your graph, ensure the sensitive elements have a visibility property and then set this visibilityProperty field to that property name. This ensures Gaffer knows to restrict access to sensitive elements.
```
Comment (cn337131) [inline docs/getting-started/advanced-guide/schema.md]: I would potentially swap the sentence order, you usually would want the quick overview of what the property is before the explanation of how to implement it.

File: docs/getting-started/advanced-guide/schema.md [hunk 5]
```diff
@@ -0,0 +1,429 @@
...
+- `timestampProperty` - When using the timestamp property in your graph, then set this timestampProperty field to that property name, so Gaffer Stores know to treat that property specially. This property allows Store implementations like Accumulo to optimise the way the timestamp property is persisted. For these stores using it can have a very slight performance improvement due to the lazy loading of properties. Setting the timestampProperty is optional and does not affect the queries available to users. For more information [see the timestamp section of the Accumulo Store Reference](../../reference/stores-guide/accumulo.md#timestamp).
```
Comment (cn337131) [inline docs/getting-started/advanced-guide/schema.md:49]: I feel like this is missing a really simple explanation of what the timestamp property is.

File: docs/getting-started/advanced-guide/schema.md [hunk 6]
```diff
@@ -0,0 +1,429 @@
...
+
+## Java API
+
+Schemas can be loaded from a JSON file directly using the [`fromJSON()` method of the `Schema` class](https://gchq.github.io/Gaffer/uk/gov/gchq/gaffer/store/schema/Schema.html#fromJson(java.io.InputStream...)). This accepts `byte[]`, `InputStream` or `Path` types, for example:
+
+```java
+Schema mySchema = Schema.fromJson(Paths.get("path/to/schema.json"));
+```
+
+While it's easiest to create Schemas objects using the JSON approach, you can create them directly from their component classes in Java.
```
Comment (cn337131) [inline docs/getting-started/advanced-guide/schema.md]:
```suggestion
While it's easiest to create Schema objects using the JSON approach, you can create them directly from their component classes in Java.
```

File: docs/getting-started/advanced-guide/schema.md (hunk 3 above)
Comment (GCHQDeveloper314) [reply docs/getting-started/advanced-guide/schema.md]: I'm not sure about this. I added lines 38 & 39 to provide an example, but this example is separate from the main example schema discussed on the page. Which is why I wrote 'that' rather than 'this'. Maybe it would be better to rephrase in general?

File: docs/getting-started/advanced-guide/schema.md (hunk 1 above)
Comment (GCHQDeveloper314) [reply docs/getting-started/advanced-guide/schema.md]: I've reworded this.

File: docs/getting-started/advanced-guide/schema.md (hunk 4 above)
Comment (GCHQDeveloper314) [reply docs/getting-started/advanced-guide/schema.md]: I've reworded/reordered this. Should now be clearer what each property is first, before going into more detailed.

File: docs/dev/components/graph.md [hunk 7]
```diff
@@ -7,16 +7,31 @@ It provides users with a single point of entry for executing operations on a sto
 
 When you instantiate a `Graph`, this doesn't mean you are creating an entirely new graph with its own data, you are simply creating a connection to a store where some data is held.
 
+See the [Graph Javadoc](https://gchq.github.io/Gaffer/uk/gov/gchq/gaffer/graph/package-summary.html) for further documentation.
+
+## Creating a Graph
```
Comment (tb06904) [inline docs/dev/components/graph.md:12]: Is it worth giving an example of connecting to an existing graph via Java? e.g. like how you would via the `gafferpy` API using the gaffer connector.

File: docs/getting-started/advanced-guide/schema.md [hunk 8]
```diff
@@ -0,0 +1,430 @@
...
+
+These 2 optional fields are for advanced users. They can go in the Elements Schema, however we have split them out into separate Validation and Aggregation Schema files for this page, so the logic doesn't complicate the Elements schema.
+
+- `validateFunctions` - An array of selections and predicates to be applied to an element. This allows you to validate based on multiple properties at once. E.g. check a timestamp property together with a time to live property to check if the element should be aged off. Individual property validation is best done as a validateFunction in the property type definition in Types schema.
+- `aggregateFunctions` - An array of selections and binary operators to be applied to an element. This allows you to aggregate based on multiple properties at once. It is important to note that types of properties (groupBy, non-groupBy, visibility) cannot be aggregated in the same aggregate function. The timestamp property is treated as a non-groupBy property. Individual property aggregation is best done as a aggregateFunction in the property type definition in the Types schema.
+
+### Example Elements Schema
+Here is an example Elements schema:
```
Comment (tb06904) [inline docs/getting-started/advanced-guide/schema.md]: Do we still want to use the road traffic example here? It's quite difficult to follow and haven't really explained what a graph with this schema would look like or how it is used

File: docs/dev/components/graph.md (hunk 7 above)
Comment (GCHQDeveloper314) [reply docs/dev/components/graph.md:12]: You would do this using the proxy store. I'll have a look at adding something around this.

File: docs/getting-started/advanced-guide/schema.md (hunk 8 above)
Comment (GCHQDeveloper314) [reply docs/getting-started/advanced-guide/schema.md]: I took it directly from the old docs. I'll have a look at simplifying it or explaining the context.

Comment (tb06904) [reply docs/getting-started/advanced-guide/schema.md]: Fair that's understandable, I wasn't sure if it was in scope to come up with another example but maybe just some context around it would help

File: docs/getting-started/advanced-guide/schema.md (hunk 3 above)
Comment (cn337131) [reply docs/getting-started/advanced-guide/schema.md]: I had another read and I am still not sure - I think maybe rephrase? I feel the 'that' makes it less clear that you mean the example in the previous sentence.

Comment (GCHQDeveloper314) [reply docs/getting-started/advanced-guide/schema.md]: I've reworded this to refer to the previous example to avoid any confusion.

File: docs/dev/components/graph.md (hunk 7 above)
Comment (GCHQDeveloper314) [reply docs/dev/components/graph.md:12]: I've added a section below this linking to the Proxy Store page. It helpfully already has an example for how to do this.

File: docs/getting-started/advanced-guide/schema.md (hunk 8 above)
Comment (GCHQDeveloper314) [reply docs/getting-started/advanced-guide/schema.md]: I've added some context and a link to where the example originates.

File: docs/getting-started/advanced-guide/schema.md [hunk 9]
```diff
@@ -0,0 +1,430 @@
...
+When using Java directly, you don't need to use JSON files to store you schemas - although this is still recommended. See the [Java API](#java-api) section for more about interacting with Schemas in Java.
+
+The sections below walkthrough the features of Schemas in detail and explain how to accurately model your data in Gaffer.
+
+
+## Elements schema
+The Elements schema is designed to be a high level document describing what information your Graph contains, i.e. the different kinds of edges and entities and the list of properties associated with each.
+Essentially this part of the schema should just be a list of all the entities and edges in the graph. 
+Edges describe the relationship between a source vertex and a destination vertex. 
+Entities describe a vertex. We use the term "element" to mean either and edge or and entity.
```
Comment (cn337131) [inline docs/getting-started/advanced-guide/schema.md]:
```suggestion
Entities describe a vertex. Edges describe the relationship between a source vertex and a destination vertex.
We use the term "element" to mean either an edge or an entity.
```

I would swap these just so it flows a little better and also just fixed the grammar.

File: docs/getting-started/advanced-guide/schema.md [hunk 10]
```diff
@@ -0,0 +1,430 @@
...
+Groups must be completely unique and cannot be shared between edges and entities.
+
+### Edges
+Edges must have the following:
+
+- `source` - Type of object to use as the source vertex in your graph. Needs to be a type defined in the [Types schema](#types).
+- `destination` - Type of object to use as the destination vertex in your graph. Can either be the same type as `source`, or a different type.
+- `directed` - Tells Gaffer if the edge is directed or undirected. Needs to be a type which means true or false, see [Types](#true--false) for more info.
+
+When an Edge is undirected in Gaffer (`directed` is `false`), it is treated as if the relationship was bidirectional, and the vertices of the edge do not have an authoritative source and destination.
```
Comment (cn337131) [inline docs/getting-started/advanced-guide/schema.md]:
```suggestion
When an Edge is undirected in Gaffer (`directed` is `false`), it is treated as if the relationship was bidirectional, meaning that the vertices of the edge do not have an authoritative source and destination.
```

File: docs/getting-started/advanced-guide/schema.md [hunk 11]
```diff
@@ -0,0 +1,430 @@
...
+Thus, the undirected edges `A -- B` and `B -- A` are equal, and will therefore be aggregated together.
+Gaffer will present the undirected edges Vertices in natural ordering, so a client will also see the above edge presented as `A, B`.
```
Comment (cn337131) [inline docs/getting-started/advanced-guide/schema.md]:
```suggestion
Gaffer will present the undirected edges vertices in natural ordering, so a client will also see the above edge presented as `A, B`.
```

File: docs/getting-started/advanced-guide/schema.md [hunk 12]
```diff
@@ -0,0 +1,430 @@
...
+It will be aggregated with any other edge that has the same source and destination and is also undirected.
+For example an undirected Edge that is added, `A -> B`, will be aggregated with another Edge of `B -> A` if they are undirected.
```
Comment (cn337131) [inline docs/getting-started/advanced-guide/schema.md]: I feel like this paragraph is a bit repetitive? Instead I would combine a couple of sentences:
Thus, the undirected edges A -- B and B -- A are equal, and will therefore be aggregated with any other edge that has the same source and destination and is also undirected. Gaffer will present the undirected edges vertices in natural ordering, so a client will also see the above edge presented as A, B. This would mean that an undirected Edge that is added, A -- B, will be aggregated with another Edge of B -- A if they are undirected.

Comment (cn337131) [reply docs/getting-started/advanced-guide/schema.md]: I also wasn't sure if the two bottom two examples should be -- instead of -> as they are meant to be undirected

File: docs/getting-started/advanced-guide/schema.md [hunk 13]
```diff
@@ -0,0 +1,430 @@
+# Schemas
+This page delves into more detail on Schemas. As seen in the other guides, there are two main components to schemas: Elements and Types.
+
+Schemas are often separated into two JSON files (`elements.json` & `types.json`), but it's worth remembering that your schema can be broken up further if required.
+
+For example if you have a Graph that is made up of 5 different data sources, you may find it easier to develop and maintain by splitting your schema into multiple parts, for example 5 Element Schemas and 1 Types schema.
+Alternatively, a single overall schema file could be used for Elements and Types. There's also no requirement to have a single Types schemas.
```
Comment (cn337131) [inline docs/getting-started/advanced-guide/schema.md]:
```suggestion
Alternatively, a single overall schema file could be used for Elements and Types, as there is no requirement to have a single Types schemas.
```

File: docs/getting-started/advanced-guide/schema.md (hunk 12 above)
Comment (GCHQDeveloper314) [reply docs/getting-started/advanced-guide/schema.md]: This section could certainly be more concise, I'll reword.

Comment (GCHQDeveloper314) [reply docs/getting-started/advanced-guide/schema.md]: I've make it more like what's suggested and also trimmed it further. That last sentence is probably not needed but it does make things very clear.

**Review by tb06904 — APPROVED** [review]
